### PR TITLE
4-bit QLoRA via bitsandbytes (4-bit base model + LoRA)

### PR DESCRIPTION
**Pull request by TimDettmers**
This adds QLoRA support to PEFT. More information about QLoRA from our abstract:

>We develop QLoRA tuning, a method that finetunes by backpropagating gradients through a frozen 4-bit base model into low rank adapters (LoRA). With QLoRA tuning we can finetune 30B/65B parameter models on 24/48GB GPUs while preserving regular 16-bit full finetuning runtime and task performance. We achieve the memory efficiency and quantization precision through a combination of new methods: nested quantization to reduce the average memory footprint from 4.5 to 4.1 bits per parameter, paged optimizers to manage gradient checkpointing memory spikes, and a new data type, 4-bit NormalFloat (NF4), which is information theoretically and empirically optimal for normally distributed weights. To demonstrate the effectiveness and ease of use of QLoRA tuning we finetune more than 1,000 models to create a detailed dissection of instruction following performance across datasets (FLAN, Alpaca, Chip2, SuperNatural Instructions, Chip2, AnthropicHH), models types (LLaMA, T5), and model scales (125M to 65B). A discussion of the results is forthcoming in our paper.

@sgugger @younesbelkada

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -41,7 +41,6 @@
         "packaging>=20.0",
         "psutil",
         "pyyaml",
-        "torch>=1.13.0",
```
Comment (younesbelkada) [inline setup.py:44]: I think this change is not needed

File: examples/int8_training/peft_adalora_whisper_large_training.py [hunk 2]
```diff
@@ -18,7 +18,6 @@
 import numpy as np
 import torch
 import transformers
-import wandb
```
Comment (younesbelkada) [inline examples/int8_training/peft_adalora_whisper_large_training.py:21]: Probably this change is not needed either

File: src/peft/peft_model.py [hunk 3]
```diff
@@ -734,6 +734,10 @@ def forward(
     def generate(self, **kwargs):
         peft_config = self.active_peft_config
         self.base_model.prepare_inputs_for_generation = self.prepare_inputs_for_generation
+        if hasattr(self.base_model, "model"):
+            self.base_model.model.generation_config = self.generation_config
+        else:
+            self.base_model.generation_config = self.generation_config
```
Comment (younesbelkada) [inline src/peft/peft_model.py:740]: I think these changes are fine for now, I will investigate later why this change is needed

**Review by younesbelkada — COMMENTED** [review]
Thank you for your insipiring work, as always! I just left 3 comments - also curious to see what @sgugger will say!

**Review by sgugger — APPROVED** [review]
Thanks for adding the support to 4bit quantization! Apart from Younes comments on potentially unrelated changes, this looks great to me!

**Comment by TimDettmers** [comment]
Thank you, Younes & Sylvain! Thanks for the changes, Younes. This looks all good to me.

**Review by younesbelkada — APPROVED** [review]
Again thank you for your great work, Tim!

**Comment by ewof** [comment]
holy shit it's happening

**Comment by ewof** [comment]
dettmers is a hero
